### ISSUE TITLE
Factor a prettier printer into separate library

### ISSUE THREAD

**Issue by Gozala** [prettier/prettier#2029]
Hi,

Recently I have being playing on library for pretty printing JS values, somewhat inspired by  [Rust Debug](https://doc.rust-lang.org/stable/std/fmt/trait.Debug.html) trait. This experiment reminded me a conversation with @jlongster and that this project  was using Philip Wadler's A prettier printer implementation for pretty printing. I would really like if the pretty printer implementation could be factored out into separate library, so that I could use it for pretty printing JS values.

I did look at the implementation in this repo and paper and it seems that implementation here somewhat diverged presumably to accommodate additional use cases. Apart form that, this implementation had being somewhat confusing to me. The reason I'm pointing out this is, I'd be willing to do actual work to do factoring, documenting and cleaning up API as long as someone who knows understands current implementation can walk me through it.

I would also really like the prettier-printer implementation to be in typescript and I'm willing to do the necessary changes if there will be a support for this.

**Comment by vjeux** [comment]
The raw printer that is based on Wadler paper is actually pretty well separated. It contained in those four files with no dependencies:

![image](https://user-images.githubusercontent.com/197597/26892648-d3dc07ce-4b6d-11e7-9e3e-17cb6f023146.png)

At this point it's been stable so it wouldn't be annoying to extract it out.

I'd be down if you want to move it to prettier/prettier-printer (or some other name) and make this project use it as a dependency.

As for moving it to TypeScript, I would rather not. So far we've been able to get the project run without any transpilation and it's been really useful for getting people to hack on it. But if you want to write a .d.ts file for it, that would be great.

**Comment by Gozala** [comment]
> The raw printer that is based on Wadler paper is actually pretty well separated. It contained in those four files with no dependencies:

Yep those where the files I was looking at.

> At this point it's been stable so it wouldn't be annoying to extract it out.

> I'd be down if you want to move it to prettier/prettier-printer (or some other name) and make this project use it as a dependency.

Perfect that's exactly what I wanted to do!

> As for moving it to TypeScript, I would rather not. So far we've been able to get the project run without any transpilation and it's been really useful for getting people to hack on it. But if you want to write a .d.ts file for it, that would be great.

Understood! How about making tiny changes to make it more type friendly. `Doc` type representation could be more formal.

I also really need some help to get my head around the implementation details, especially where it diverges from the paper.

**Comment by vjeux** [comment]
> How about making tiny changes to make it more type friendly

Yeah, feel free to change the API as long as you also fix it inside of prettier.

**Comment by vjeux** [comment]
I just added you as an admin to https://github.com/prettier/prettier-printer, go nuts! :)

For the implementation details, @karl and @jlongster are likely your best candidates.

**Comment by Gozala** [comment]
Thanks! It also turns out https://github.com/prettier/prettier/blob/master/commands.md does have answers to my questions.

**Comment by vjeux** [comment]
I just added all the commands that were not documented: https://github.com/prettier/prettier/commit/e1ebbff001cc6f30398d39ec15cc419b3ece66ec

**Comment by Gozala** [comment]
@vjeux thanks I have pushed printer related files to https://github.com/prettier/prettier-printer. In an initial version I can just keep everything as is and just update this repo to use prettier/prettier-printer as a dependency.

**Comment by Gozala** [comment]
I'll post some questions here as I'll run into them:

1. `softline` it seems that it can be expressed as `ifBreak(line, '')` no ? If so, is there any reason no to change it to exactly that ?
2. Is `breakParent` just an implementation detail of `hardline` and `literalline` or is it useful on it's own ? Or to put it differently is there a scenario where `breakParent` could be present and not be followed by `line`, what is such scenario ? 
3. Here is the definition for `const hardline = concat([{ type: "line", hard: true }, breakParent]);` and I'm not sure what purpose does `hard:true` serves. Is not a `breakParent` here that actually forces the breaking ?
4. Somewhat related to above, [is this condition redundant](https://github.com/prettier/prettier/blob/f702c2c61484241f1d0ca33a6603dcdcc0102256/src/doc-utils.js#L103-L105) it seems that next condition would always capture this.
5. What the purpose of [`placeholder` filed in `cursor`](https://github.com/prettier/prettier/blob/f702c2c61484241f1d0ca33a6603dcdcc0102256/src/doc-builders.js#L88):
    ```js
    const cursor = { type: "cursor", placeholder: Symbol() };
    ```

**Comment by vjeux** [comment]
1. I've also never understood why. I believe that

```js
line = ifBreak('\n', '')
softline = ifBreak('\n', ' ')
hardline = '\n'
```

I think it makes sense to have those as a high level API, but I don't know why they need some special implementation. Also, I really hate the name `line` vs `softline`. I think that `line_or_nothing` and `line_or_space` would be much more explicit.

2. We have two call sites of breakParent: 

- `printed.some(willBreak) ? breakParent : "",`
- `willBreak(oneLine) ? breakParent : "",`

I don't know for sure if they are there because we can't use `group(..., { shouldBreak: true })` or if they can get rid of. I would just try it and see what tests break. If none break, then ship it :)

3. No idea, @jlongster?

4. Same thing, try to remove it and run the tests. If they pass, you can get rid of it.

5. @josephfrazier ?

**Comment by jlongster** [comment]
> softline it seems that it can be expressed as ifBreak(line, '') no ? If so, is there any reason no to change it to exactly that ?

Newlines are very special in prettier and if you look at the low-level printer it does a lot of special things when it hits a line. They aren't just dump strings to print to the screen, they are critical for measuring.

> Here is the definition for const hardline = concat([{ type: "line", hard: true }, breakParent]); and I'm not sure what purpose does hard:true serves. Is not a breakParent here that actually forces the breaking ?

Same reason as above, lines are very different and `breakParent` is what forcefully breaks all parents, but we still need to mark it as a hard line because it changes measuring. Search for `shouldRemeasure` in doc-printer.js.

Some of this might be able to be written better though. The low level printer could use some love and it could be more straight-forward, but newlines are definitely an integral part of it.

**Comment by jlongster** [comment]
This has diverged a ton from the paper, or at least a ton has been added. As Brandon Bloom [said](https://twitter.com/BrandonBloom/status/867222817577762816), there are two very different use cases for pretty printing: code and data. It sounds like you want the former, and I don't think you'll need a lot that we have here.

The core algorithm is very simple. The really annoying parts are the when we want to do something "special", like breaking an inner group before an outer group (see how we expand an object that is the last argument of a function). You never need to do that with data structures.

Personally I think you're better off learning from this but implementing your own and adding whatever primitives you need. You'll have more freedom to add whatever you want, which I find is helpful when you're trying to format something in a specific way. And selfishly, if we spin it off, it's another thing we need to maintain and generalize and can't change easily in the future.

At the very core though, it's just `group`, `concat`, and `line` I think. And it's a simple recursive walk to do to the measuring and printing (separately). You could implement the first version in probably 30 lines, and I'd be happy to help with that.

I have been wanting to spend time cleaning this up so it's more readable. We manually flatten the recursive loop into an imperative one, but I don't think that's actually needed (copied that from a Rust implementation). There is also at least one pretty serious bug that I've been meaning to fix that we've been able to sidestep.

Basically, I'm not opposed to it, but with all the activity this project gets, it's really nice for it to be internal and not worry about changing things :) But I understand wanting to share code. I think it'd be worthwhile to at least start with your own copy and then we can see if there's a core thing that we can all share.

**Comment by Gozala** [comment]
> softline it seems that it can be expressed as ifBreak(line, '') no ? If so, is there any reason no to change it to exactly that ?
> 
> Newlines are very special in prettier and if you look at the low-level printer it does a lot of special things when it hits a line. They aren't just dump strings to print to the screen, they are critical for measuring.
> 
>     Here is the definition for const hardline = concat([{ type: "line", hard: true }, breakParent]); and I'm not sure what purpose does hard:true serves. Is not a breakParent here that actually forces the breaking ?
> 
> Same reason as above, lines are very different and breakParent is what forcefully breaks all parents, but we still need to mark it as a hard line because it changes measuring. Search for shouldRemeasure in doc-printer.js.
> 
> Some of this might be able to be written better though. The low level printer could use some love and it could be more straight-forward, but newlines are definitely an integral part of it.

Thanks for explaining, I have probably overlooked `shouldRemeasure` will look at it get a clearer picture.

**Comment by Gozala** [comment]
> This has diverged a ton from the paper, or at least a ton has been added. As Brandon Bloom said, there are two very different use cases for pretty printing: code and data. It sounds like you want the latter (whoops, edited this, said former before), and I don't think you'll need a lot that we have here.

Thanks for the pointer & linked paper also seems relevant and interesting :) You are right the mostly what I want is data, but it also includes printing functions (not gonna go into details here) which pretty much means I need former as well. I have considered printing just function signatures instead, but then I realized I could probably reuse your work instead of compromising :)

> Personally I think you're better off learning from this but implementing your own and adding whatever primitives you need. You'll have more freedom to add whatever you want, which I find is helpful when you're trying to format something in a specific way. And selfishly, if we spin it off, it's another thing we need to maintain and generalize and can't change easily in the future.

I don't necessarily mind starting from scratch, in fact I have a semi-complete implementation after that paper. But then I thought collaborating on the existing code base would be a better option as maintenance burden would be shared, if pretty-printer happens to have things I don't care, I can just simply not use them.

> Basically, I'm not opposed to it, but with all the activity this project gets, it's really nice for it to be internal and not worry about changing things :) But I understand wanting to share code. I think it'd be worthwhile to at least start with your own copy and then we can see if there's a core thing that we can all share.

After talking to @vjeux yesterday I end up moving printer to https://github.com/prettier/prettier-printer and was planning to send a pull request here replacing to remove `doc-*.js` files and use a dependency `prettier-printer` dependency instead. After that I was going to attempt to clean up a code base a bit & simplify where possible, which is why I have being trying to get better understanding of why things are the way the are.

@jlongster After reading your last comment I'm not sure what the official position on this is. Should I drop this effort or should I carry on as planned ? Or maybe do some course correction ?

**Comment by Gozala** [comment]
I have being trying to write tests for `breakParent` but I clearly don't quite understand it, so here is an example:

```js
const doc = group(
    concat([
      "abcdef", // 1...6
      line, // 7
      group(
        concat([
          "ghi", // 8..10
          line, // 11
          breakParent,
          "jk", // 12..13
          line, // 14
          "l"   // 15
        ])
      ),
      line, // 16
      "mnopq", // 17..21
    ])
  )
```

I would expect that with `printWidth:13` only outer group will have line breaks while inner group will have spaces: `"abcdef\nghi jk l\nmnopq" given that the whole thing does not fit on one line while inner group does fit on a single line. Can someone please elaborate what does expected behavior should actually be ?

**Comment by vjeux** [comment]
> if group contains breakParent and that group does not fit it will cause all parents to switch to break mode

If a group contains breakParent, the group itself will be forced to break and that break will propagate up to all the parents.

The parent here means "the nearest parent group", so that's the group you are talking about, not its parent.

**Comment by josephfrazier** [comment]
> 5. What the purpose of [`placeholder` filed in `cursor`](https://github.com/prettier/prettier/blob/f702c2c61484241f1d0ca33a6603dcdcc0102256/src/doc-builders.js#L88):
    ```
    const cursor = { type: "cursor", placeholder: Symbol() };
    ```

It is [added](https://github.com/prettier/prettier/blob/09f147decd17cc2c95286d19ba0d6801a4b9a4e9/src/comments.js#L936-L941) to the document (then to the array of strings constituting the output [here](https://github.com/prettier/prettier/blob/7114eb4f1e5efc99f6125f0ea187f5458c979939/src/doc-printer.js#L160)), so that the cursor position after formatting can be [calculated](https://github.com/prettier/prettier/blob/7114eb4f1e5efc99f6125f0ea187f5458c979939/src/doc-printer.js#L417-L426). I used a Symbol because it's guaranteed to be a unique value. See https://github.com/prettier/prettier/pull/1637 for more context.

**Comment by Gozala** [comment]
> If a group contains breakParent, the group itself will be forced to break and that break will propagate up to all the parents.
> 
> The parent here means "the nearest parent group", so that's the group you are talking about, not its parent.

Thanks for explanation, I was confused by the naming, so what it actually means is force break this group and all of it's and propagate to all parent groups.

**Comment by Gozala** [comment]
> It is added to the document (then to the array of strings constituting the output here), so that the cursor position after formatting can be calculated. I used a Symbol because it's guaranteed to be a unique value. See #1637 for more context.

@josephfrazier Thanks for explaining! I think you can just use `cursor` object itself as unique value, which is kind of why I was confused ?

**Comment by josephfrazier** [comment]
> @josephfrazier Thanks for explaining! I think you can just use cursor object itself as unique value, which is kind of why I was confused ?

Oh, yeah, I guess we could just use the object itself. When I was developing this feature, I had originally thought that everything in the `out` array needed to be a string, so I used a random string as the placeholder value. Later, I realized I could use other data types, so I switched to a Symbol, but  it didn't occur to me to just use the object and get rid of the `placeholder` property.

**Comment by Gozala** [comment]
Since I have not heard back from @jlongster yet and it seemed like he was inclined towards not factoring out printer, I end up implementing printer that follows paper one to one.

https://gist.github.com/Gozala/3841ea4920113fdca31ecbf5a4e5b803

I'm still curious if additional features like `softline`, `hardline`, `literalline` and others can be added without changing it too much. But unfortunately I don't quite understand required invariant mentioned in paper:

> Again, as an invariant, we require in `(x :<|> y)` that all layouts in `x` and `y` flatten to the same layout, and that no first line in `x` is shorter than any first line in `y`.
>```hs
> x <+/> y = x <> (text " " :<|> line) <> y
> fillwords = folddoc (<+/>) . map text . words
>```
> Recall that we do not expose `:<|>` to the user, but `x <+/> y` may be safely ex- posed, because it satisfies the invariant required by `:<|>`. Both `text " "` and line flatten to the same layout, and the former has a longer first line than the latter.

I really don't understand how `text " "` has a longer first line than `line` they seem to have equal length from what I can tell. Posting it here as @jlongster might have insight.

**Comment by Gozala** [comment]
Just a quick update:

1. @alexandrebodin had expressed interest in helping with this on gitter. I think we missed each other so mentioning him here to get him into a loop.
2. **Current status**: Code has being extracted to [prettier-printer](https://github.com/prettier/prettier-printer). It has very few tests as there were none in this repo. I started writing some, but then I was not entirely sure how certain constructs supposed to behave (as can you can tell from my questions posted earlier).
3. It is no longer clear where things stand in terms of supporting this effort. As far as I understand it is undesired to use factored out printer as a dependency here. There for I don't plan on creating a pull request doing that.
4. It is unclear to me where things stand in terms of publishing [prettier-printer](https://github.com/prettier/prettier-printer) to npm. If it's 👍 then I feel like there should be at least enough test coverage to illustrate each doc builder function, even if it does not capture all edge cases. This could use some help.
5. Given the way things stand now, I have no initiative to pursue this effort. Truth is, my needs were well enough served with a pretty printer in paper, all the extensions made here were nice to have in case I need them. Mostly I was hoping to share maintenance efforts for pretty-printer library, but given an unpopularity of this idea, I think I'd be better of with an implementation that in direct implementation after pretty printer paper & that I have complete understanding off. In other words unless there is interest in factoring printer out and than collaborating on it together, I don't really feel like committing to [prettier-printer](https://github.com/prettier/prettier-printer). 

I hope my position is reasonable & thanks to everyone who helped me out by answering questions.

**Comment by vjeux** [comment]
Since you've gone through the work of making the internal lib standalone, you should push it to npm at least, this way other people can use it. This has been a recurrent request so I'm sure it's not going to be useless :)

**Comment by azz** [comment]
Any update on this?

**Comment by vjeux** [comment]
@Gozala seems no longer interested in pursuing this work, so let's close this one out.
